### PR TITLE
Adding support for arm64 images

### LINKED ISSUE #1200: Arm32v6/7 or Arm64v8 for docker?
Is it possible to use an arm machine for mailu?  I've looked through the tags and there doesn't seem to be a specific arm tag, and I can't see if a manifest has been made.

If not, are there build docker instructions so I could build the images myself?

### PR DESCRIPTION
**Pull request by ofthesun9**
## What type of PR?
Enhancement & bug-fix (fix #1200)

## What does this PR do?
This PR adds support for arm64 architecture, thanks to Travis multi-platform capabilities.
Tests are done on both amd64 and arm64 architectures, and manifest(s) are created with the same image naming than current naming, so that the PR is transparent for the existing users, and no change is needed on the setup/flavor side.

### Related issue(s)
Fix #1200 

## Prerequistes
- [X] In case of feature or enhancement: documentation updated accordingly
- [X] Unless it's docs or a minor change: add [changelog](https://mailu.io/master/contributors/guide.html#changelog) entry file.

### CODE REVIEW

**Comment by ofthesun9** [comment]
bors try

**Comment by ofthesun9** [comment]
bors try

**Comment by ofthesun9** [comment]
Seems that clamav is currently  broken on arm64 alpine:3.10 & 3.11
Works ok on alpine:edge

**Comment by muhlemmer** [comment]
Alpine latest was recently bumped to 3.11.6. Fingers crossed :crossed_fingers: 

bors try

**Comment by muhlemmer** [comment]
bors try

**Comment by muhlemmer** [comment]
Nothing is running on travis. Highly strange...

bors try

**Comment by muhlemmer** [comment]
bors try-

**Comment by muhlemmer** [comment]
bors try

**Comment by ofthesun9** [comment]
I filled a PR an [alpine](https://gitlab.alpinelinux.org/alpine/aports/issues/11340) :
rspamd-2.5 is working Ok (running here since several weeks now!) but is available only on alpine edge.

I can provide a [commit](https://github.com/ofthesun9/Mailu/commit/890dec7b3be1c5fe876abb93da82955d3c007e1f) to use the edge repository for the antispam container, but up to now I was waiting for a fix that would allow us to stick to an alpine "table" release.
Now, I realize that it could take several months... 

So if we all agree to have this temporary fix for the container (until alpine 3.12 is released), I can push it.

**Comment by muhlemmer** [comment]
My proposal would be to set the `DISTRO` environment variable for ARM to the `alpine:edge` image:

https://github.com/Mailu/Mailu/blob/15a0d7303c70ac17ea3061651a5fdcabd4c69737/core/rspamd/Dockerfile#L1-L2

This way we don't impact the stable releases on amd64. I'm not sure when alpine 3.12 will ship and if it will be before Mailu 1.8.

**Comment by ofthesun9** [comment]
bors try

**Comment by ofthesun9** [comment]
bors try

**Comment by ofthesun9** [comment]
The images related to this PR can be tested through:
docker-compose down
DOCKER_ORG=mailutest MAILU_VERSION=pr-1421 docker-compose up -d

**Comment by ofthesun9** [comment]
I changed antispam Dockerfile to use alpine:3.12 (and rspamd 2.5).
bors try

**Comment by elFarto** [comment]
I've tried deploying this to my Raspberry PI 3B, running 64-bit Raspberry PI OS, and the antispam container is failing to start. Stepping into the container and starting it manually results in this:

```
bash-5.0# ./start.py 
Segmentation fault (core dumped)
```

After some fiddling, I managed to get a backtrace from the core dump:
```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x0000007f9b246908 in lpeg_allocate_mem_low (sz=<optimized out>, sz@entry=512) at /usr/include/fortify/string.h:53
```

There appears to be a similar issue and fix (but for ppc64el not arm64) here: rspamd/rspamd#3365. It doesn't look like there's been a release since that fix was merged.

**Comment by ofthesun9** [comment]
@elFarto Thanks for the feedback :-) I already experienced crashes at start-up, and it seems that the host OS is important here (some images will run on debian and not on ubuntu and vice versa...)

I was hoping that images build on alpine:3.12 would solve it (at least the ubuntu OS on travis is now Ok :-) ) but your message shows that it is still not ok on some host OS :-(

AFAIK, the images from this PR run Ok on ubuntu 16.04, 18.04 and 20.04

I have some image built with luajit disabled: could you please give it a try ?:
ofthesun9/rspamd:test-debian-multiarch-rspamdgitbuild

**Comment by elFarto** [comment]
That new rspamd image starts up correctly.

File: tests/deploy_manifest.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+fi
+
+docker login -u $DOCKER_UN -p $DOCKER_PW
+
+# List of all Mailu services:
+MailuSrvList="nginx unbound dovecot postfix rspamd clamav radicale traefik-certdumper admin postgresql roundcube rainloop fetchmail none docs setup"
+# Iterate for each Mailu service. Retag images (removing -ci), and create multiarch manifest  
+for MailuSrv in $MailuSrvList; do
+    echo "Retagging ${MailuSrv} and pushing it, for amd64 and arm64"
+    docker pull ${DOCKER_ORG:-mailu}/${DOCKER_PREFIX:-}${MailuSrv}${MAILU_CI}:amd64-${MAILU_VERSION:-local}
```
Comment (Yajo) [inline tests/deploy_manifest.sh:21]: IMHO it's better to push a multiarch tag, instead of arch-specific ones.

File: tests/deploy_manifest.sh [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    echo "Creating manifest for ${MailuSrv}"
+    DOCKER_CLI_EXPERIMENTAL=enabled  docker manifest create \
+       ${DOCKER_ORG:-mailu}/${DOCKER_PREFIX:-}${MailuSrv}:${MAILU_VERSION:-local} \
+       --amend  ${DOCKER_ORG:-mailu}/${DOCKER_PREFIX:-}${MailuSrv}:amd64-${MAILU_VERSION:-local} \
+       --amend  ${DOCKER_ORG:-mailu}/${DOCKER_PREFIX:-}${MailuSrv}:arm64-${MAILU_VERSION:-local}	
+    DOCKER_CLI_EXPERIMENTAL=enabled  docker manifest annotate \
+       ${DOCKER_ORG:-mailu}/${DOCKER_PREFIX:-}${MailuSrv}:${MAILU_VERSION:-local} \
+       ${DOCKER_ORG:-mailu}/${DOCKER_PREFIX:-}${MailuSrv}:arm64-${MAILU_VERSION:-local} --os linux --arch arm64 --variant v8
+    DOCKER_CLI_EXPERIMENTAL=enabled  docker manifest push \
+       ${DOCKER_ORG:-mailu}/${DOCKER_PREFIX:-}${MailuSrv}:${MAILU_VERSION:-local}
```
Comment (ofthesun9) [inline tests/deploy_manifest.sh:38]: the multiarch manifest is created here, just after all arch specific images have been pushed

File: .travis.yml [hunk 3]
```diff
@@ -13,42 +13,108 @@ branches:
...
-  - docker-compose -v
-  - docker-compose -f tests/build.yml build
-  - sudo -- sh -c 'mkdir -p /mailu && cp -r tests/certs /mailu && chmod 600 /mailu/certs/*'
 
-
-script:
+jobs:
+  include:
+# install stage will build and push 
+# all images (names will get the suffix defined bu MAIL_CI)
```
Comment (HorayNarea) [inline .travis.yml:33]:
```suggestion
# all images (names will get the suffix defined bu MAILU_CI)
```

**Comment by Nebukadneza** [comment]
This PR looks well-developed, and nicely done, and @ofthesun9 uses it successfully.
Despite this, considering the current time-invest issues among the maintainers, we unfortunately decided to move this to the next release. This was in the hope that, after 1.8, we can target a small and rather early followup-release.

**Comment by ofthesun9** [comment]
bors try

**Comment by ofthesun9** [comment]
bors try

**Comment by ofthesun9** [comment]
bors try

**Comment by ofthesun9** [comment]
bors try

**Comment by ofthesun9** [comment]
bors try

**Comment by ofthesun9** [comment]
try

**Comment by ofthesun9** [comment]
bors try

**Comment by ofthesun9** [comment]
bors try

**Comment by ofthesun9** [comment]
Bors try

**Comment by armandleopold** [comment]
Deploying on a 4x Raspberry Pi 4 4Go K3s Cluster.
With Helm chart : [Mailu Helm chart](https://github.com/Mailu/helm-charts/blob/master/mailu/values.yaml)
The **rspamd** image doesn't work. I had to use this one : https://hub.docker.com/r/a16bitsysop/rspamd

**Comment by ofthesun9** [comment]
> Deploying on a 4x Raspberry Pi 4 4Go K3s Cluster.
> With Helm chart : [Mailu Helm chart](https://github.com/Mailu/helm-charts/blob/master/mailu/values.yaml)
> The **rspamd** image doesn't work. I had to use this one : https://hub.docker.com/r/a16bitsysop/rspamd

Which OS are you using ?

**Comment by armandleopold** [comment]
> > Deploying on a 4x Raspberry Pi 4 4Go K3s Cluster.
> > With Helm chart : [Mailu Helm chart](https://github.com/Mailu/helm-charts/blob/master/mailu/values.yaml)
> > The **rspamd** image doesn't work. I had to use this one : https://hub.docker.com/r/a16bitsysop/rspamd
> 
> Which OS are you using ?

```bash
pi@pi1:~ $ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
pi@pi1:~ $ uname -a
Linux pi1 5.4.51-v8+ #1333 SMP PREEMPT Mon Aug 10 16:58:35 BST 2020 aarch64 GNU/Linux
```

The ARM64 Raspberry OS

**Comment by ofthesun9** [comment]
@armandleopold , Ok this is the [bug](https://gitlab.alpinelinux.org/alpine/aports/-/issues/11340) I encountered in March which resurfaced...
I asked if the fix found for alpine-edge could be available in alpine 3.12

If you encounter any issues with these arm64 build, please feel free to provide feedback.

**Comment by armandleopold** [comment]
I have tested the docker images **docker.io/mailutest/*:pr-1421** and they work for me with overriding the default definition in the helm chart.
Except for the **rspamd** as mentioned above.

**Comment by ofthesun9** [comment]
bors try

**Comment by ofthesun9** [comment]
@armandleopold Could you please give the latest build a try ?. rspamd issue shoud be gone with this one...

**Comment by dgiebert** [comment]
@ofthesun9 I have deployed it using the helm chart to a k3s RPi Cluster and rspamd starts :+1:

**Comment by ofthesun9** [comment]
> @ofthesun9 I have deployed it using the helm chart to a k3s RPi Cluster and rspamd starts :+1: 

Thanks for the feedback ! Which OS are you using ?

**Comment by dgiebert** [comment]
I am using k3os

**Comment by ofthesun9** [comment]
bors try

**Comment by armandleopold** [comment]
> @armandleopold Could you please give the latest build a try ?. rspamd issue shoud be gone with this one...

I have tested. It works well now. Thanks !

**Comment by Robin-floss** [comment]
is there a way to build the images ourselves right now?

**Comment by Erriez** [comment]
@Robin-floss I am not able to build via the script `tests/build_arm.sh` locally on my AMD64 desktop. This script builds 32-bit ARM images and I need ARM64 for Raspberry Pi 4 with 8GB memory on Ubuntu server 20.04. The reason of this PR is to build these images automatically from CI server which is (probably) a non ARM target. (I assume AMD64?)

My workaround is to integrate the `tests/build.yml` `build: lines manually into my `docker-compose.yml` file for each container and build the images directly on the Pi with command `docker-compose build`. This probably works on a ARM 32-bit as well.

**Comment by nextgens** [comment]
It's too late in the cycle for it to make it to 1.8 IMHO... but there is no reason why we couldn't have something like this in master

**Comment by Erriez** [comment]
@nextgens Oh ok, good to know. Then I'll switch to `master` now.

**Comment by nextgens** [comment]
This PR is obsolete: we don't use travis anymore... I'll close it

If any of @Erriez, @Robin-floss or @ofthesun9 feel like submitting the equivalent for github actions, I'll definitely consider reviewing it in master

**Pull request closed without merging**